python/dgl/graphbolt/negative_sampler.py

```diff
@@ -0,0 +1,136 @@
+    """
+
+    def __init__(
+        self,
+        datapipe,
+        negative_ratio,
+    ):
+        """
+        Initlization for a negative sampler.
+
+        Parameters
+        ----------
+        datapipe : DataPipe
+            The datapipe.
+        negative_ratio : int
+            The proportion of negative samples to positive samples.
+        """
+        super().__init__(datapipe, self._sample)
+        assert negative_ratio > 0, "Negative_ratio should be positive Integer."
+        self.negative_ratio = negative_ratio
+
+    def _sample(self, node_pairs):
+        """
+        Generate a mix of positive and negative samples.
+
+        Parameters
+        ----------
+        node_pairs : Tuple[Tensor] or Dict[etype, Tuple[Tensor]]
+            A tuple of tensors or a dictionary represents source-destination
+            node pairs of positive edges, where positive means the edge must
+            exist in the graph.
+
+        Returns
+        -------
+        Tuple[Tensor] or Dict[etype, Tuple[Tensor]]
+            A collection of edges or a dictionary that maps etypes to edges,
+            which includes both positive and negative samples.
+        """
+        if isinstance(node_pairs, Mapping):
+            return {
+                etype: self._collate(
+                    pos_pairs, self._sample_negative_pairs(pos_pairs, etype)
+                )
+                for etype, pos_pairs in node_pairs.items()
+            }
+        else:
+            return self._collate(
+                node_pairs, self._sample_negative_pairs(node_pairs, None)
+            )
+
+    def _collate(self, pos_pairs, neg_pairs):
+        """Collates positive and negative samples.
+
+        Parameters
+        ----------
+        pos_pairs : Tuple[Tensor]
+            A tuple of tensors represents source-destination node pairs of
+            positive edges, where positive means the edge must exist in
+            the graph.
+        neg_pairs : Tuple[Tensor]
+            A tuple of tensors represents source-destination node pairs of
+            negative edges, where negative means the edge may not exist in
+            the graph.
+
+        Returns
+        -------
+        Tuple[Tensor]
+            A mixed collection of positive and negative node pairs.
+        """
+
+    def _sample_negative_pairs(self, node_pairs, etype=None):
+        """Generate negative pairs for a given etype form positive pairs
+        for a given etype.
+
+        Parameters
+        ----------
+        node_pairs : Tuple[Tensor]
+            A tuple of tensors or a dictionary represents source-destination
+            node pairs of positive edges, where positive means the edge must
+            exist in the graph.
+        etype : (str, str, str)
+            Canonical edge type.
+
+        Returns
+        -------
+        Tuple[Tensor]
+            A collection of negative node pairs.
+        """
+
+
+class IndependentNegativeSampler(NegativeSampler):
```

Comment:
If we only have 2 types of output, I feel the code will be a lot simpler if we merge them and input as parameter.

Reply:
make sense, changed for fix